src/Components/Blazor/Blazor/src/Hosting/EntrypointInvoker.cs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Blazor.Hosting
+{
+    internal static class EntrypointInvoker
+    {
+        public static Task InvokeEntrypointAsync(IntPtr entrypointMethodHandleValue, string[] args)
+        {
+            var entrypointMethodBase = CreateMethodBase(entrypointMethodHandleValue);
+
+            // For "async Task Main", the C# compiler generates a method called "<Main>"
+            // that is marked as the assembly entrypoint. Detect this case, and instead of
+            // calling "<Whatever>", call the sibling "Whatever".
+            if (entrypointMethodBase.IsSpecialName)
+            {
+                var origName = entrypointMethodBase.Name;
+                var origNameLength = origName.Length;
+                if (origNameLength > 2)
+                {
+                    var candidateMethodName = origName.Substring(1, origNameLength - 2);
+                    var candidateMethod = entrypointMethodBase.DeclaringType.GetMethod(
+                        candidateMethodName,
+                        BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic,
+                        null,
+                        entrypointMethodBase.GetParameters().Select(p => p.ParameterType).ToArray(),
+                        null);
+                    if (candidateMethod != null)
+                    {
+                        entrypointMethodBase = candidateMethod;
+                    }
+                }
+            }
+
+            // We're not handling any errors here, synchronous or asynchronous. That means they'll
+            // bubble up to the caller on the JS side, which can decide what to do with them.
+            return entrypointMethodBase.Invoke(null, new object[] { args }) as Task;
+        }
+
+        private static MethodBase CreateMethodBase(IntPtr methodHandleValue)
+        {
+            var methodHandleCtor = typeof(RuntimeMethodHandle).GetConstructor(BindingFlags.NonPublic | BindingFlags.Instance, null, new[] { typeof(IntPtr) }, null);
```

Comment:
🤒

Reply:
more reflection 🎉, more slow startup 🎉

Reply:
This gives me pause since `RuntimeMethodHandle` doesn't have any public constructors. This seems like Mono should give us an API

Reply:
> This gives me pause since RuntimeMethodHandle doesn't have any public constructors. This seems like Mono should give us an API

Reply:
> This gives me pause since RuntimeMethodHandle doesn't have any public constructors. This seems like Mono should give us an API

If this wasn't so low-level, I'd also be concerned. However I see this as being a polyfill in place of native async-main support, so use of private reflection feels reasonable and proportionate. If at some point Mono can offer this feature natively we'll switch to that, and in the meantime all we need is for this not to get broken, which I'm pretty sure will not be a problem for Mono.

Let's raise this with @lewing and get an agreement about whether (a) Mono could do this natively in the near future, or (b) the Mono team can at least be aware we're doing this and agree it's acceptable and won't break it, including when switching to CoreFx.

Reply:
> won't break it, including when switching to CoreFx.

This seems exceptionally unlikely to be remembered and dealt with 😆

Reply:
Thanks to a suggestion from @vargaz, this private reflection approach is now eliminated in favour of a regular public API solution, so I think the key problem is gone.